### ISSUE TITLE
Consider the return type in overload resolution

### ISSUE THREAD

**Issue by gusty** [fsharp/fslang-suggestions#820]
# Consider the return type in overload resolution

I propose we add a mechanism (attribute or language construct) to consider the return type in overload resolution.

The existing way of approaching this problem in F# is to add a dummy parameter with the type of the return type and call the function with its default value.

## Pros and Cons 

 - More control over overload resolution
 - Less cases leading to ambiguity resolution compile-time errors
 - Better type inference, since type inference is better when the return type is taken into account.

The disadvantages of making this adjustment to F# are:

 - Adding an extra modifier attribute or a language construct.
 - It's work.

## Extra information

Estimated cost: S-M

Related suggestions: #819 

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

**Comment by abelbraaksma** [comment]
For C# 9 I read somewhere that they're going to support return type covariance (though I think they meant contravariance, returning a more specific type than expected). It might be related, not sure, I don't fully grasp your proposal yet ;). 

@gusty for clarification, what kind of code would currently not compile, and after this change would compile successfully? Could you elaborate with some examples?

**Comment by robkuz** [comment]
@gusty I would love this feature very much. However why an additional lang construct or attribute?
Shouldnt that be the proper case in anyway?

    type Runner =
         static member Run(x: string): int = 0
         static member Run(x: string): Boolean = false

    let an_int: int = Runner.Run("0")
    let a_bool: Boolean = runner.Run("false")

**Comment by gusty** [comment]
@abelbraaksma  see @robkuz did already an example ;)

@robkuz I would love to, but that would ago against the specs (not just F# spec, also C# and possibly other langs) and more importantly it might be a breaking change.

**Comment by robkuz** [comment]
@gusty well if the Spec is wrong or nonsensical it should be tossed, shouldnt it? ;-)  
but yeah if this is the only or the fastest way ... then ...

**Comment by gusty** [comment]
Sure, let's leave it open to make it as default. Anyway I'm not 100% it would be a breaking change, because not having it normally leads to the ambiguos overload resolution error.

**Comment by charlesroddie** [comment]
Is there a use case for this? An example where it would not be better to rename the method?

I suspect that this code should probably be ruled out as a style restriction, and if it fails to compile, even better.

**Comment by gusty** [comment]
An easy example, all generic applicatives and monads require a set of overloads, based solely on the return type. See [here]( https://github.com/fsprojects/FSharpPlus/blob/33fa3411cb080fc70d8b4ad4f71e16d63bc82128/src/FSharpPlus/Functor.fs#L98-L138) an example.

A more complex scenario, many overloaded functions improve type inference if the return type is taken into account.

> I suspect that this code should probably be ruled out as a style restriction, and if it fails to compile, even better.

Hope you're not willing to forbid overloads ;)

**Comment by replicaJunction** [comment]
> Is there a use case for this? An example where it would not be better to rename the method?

I opened the issue referenced above to see if a method which currently returns `Choice<_, _>` could be changed to return `Result<_, _>`. The issue was dismissed since it would be a breaking change to the library.

If I understand this feature proposal correctly, it would allow methods that currently return variations on `Choice` to also support more explicit types, such as `Result` in this case. This would mean cleaner and more expressive code for consumers without breaking changes to existing code.

**Comment by Happypig375** [comment]
>without breaking changes to existing code.

Consider type inference. There will be ambiguous overloads.

**Comment by gusty** [comment]
@replicaJunction  to be more precise, in your specific case it would force the user to annotate the resulting type in order to solve the ambiguity, that's where it might break existing code.

I guess in most cases it would just work, even if it's not type annotated, since normally this kind of construct is followed by a match which may signal type inference that a Result is expected.

Still, even if there is a small percentage where type inference lacks information to resolve the overload, `Async.catch` is so common that the amount of code broken would be huge.

For that issue in particular, I would suggest to add another function with a different name and deprecate the existing one.

**Comment by abelbraaksma** [comment]
> I guess in most cases it would just work

@gusty, do you mean method overloads that are only overloaded on the return type? Because there's currently no way (other than reflection) to have them auto inferred. Which is correct according to the current spec, which disallows considering overloads that *only* differ in return type. 

Strangely enough, it is possible to create such methods (in C# this is forbidden, I believe). You just cannot call them.

**Comment by gusty** [comment]
@abelbraaksma I was talking on the assumption of having this suggestion in place.

> Strangely enough, it is possible to create such methods (in C# this is forbidden, I believe). 
Really? Normally the compiler complains and I need to use some tricks to make them look different.

**Comment by abelbraaksma** [comment]
@gusty (sorry for the delay), I've tried to mimic previous findings but I can't anymore, and I never saved the test prj in my repro. I know it is possible with `op_Explicit` and `op_Implicit`, since they are special, but with anything else it is not reproducible anymore.

Possibly it was just some lucky fluke with one of the many test and preview version of fsc lying around, but since I cannot recall what I did exactly, we may never know.

From a .NET perspective, I din't think there's a problem expression this in MSIL, it's a language limitation.

**Comment by gusty** [comment]
Shame @abelbraaksma  I was curious about it.

@replicaJunction some days ago after a discussion with @ninofloris he came up with a kind of pattern to define a default in absence of type annotations. That would solve the problem for `Async.catch` by making it default to Choice in absence of type annotations. But not sure how reliable is that pattern in all real world scenarios, I guess it's hard to tell.

**Comment by abelbraaksma** [comment]
> Shame @abelbraaksma I was curious about it.

haha, yeah, me too! I'm really wondering what I did back then. Who knows it'll come back at some time.

**Comment by NinoFloris** [comment]
@abelbraaksma link is here https://gist.github.com/NinoFloris/3edc906b408c991265e77f73c92a9813

**Comment by abelbraaksma** [comment]
@NinoFloris, if I read that code correctly (cursory read, granted), it allows a kind of return type polymorphism on use of the functions, but it doesn't allow direct overloads (that only differ by return type) to be declared on a class or DU, right?

**Comment by NinoFloris** [comment]
@abelbraaksma correct, as they are normal methods overload resolution won't consider their return types. Which is why the return type is also encoded as an argument on respective overloads.

**Comment by gusty** [comment]
@NinoFloris after reading your comments, I'm thinking maybe we should open another suggestion for either allow specifying a default or allow specifying that there's no default. Or maybe both of them.

That would help a lot type inference in SRTPs scenarios (which soon will become more common with the extensions visible RFC) and avoid very obscure error messages.

**Comment by vzarytovskii** [comment]
I am marking it as approved in principle, since it's in the RFC for the extension methods in SRTP: https://github.com/fsharp/fslang-design/blob/main/RFCs/FS-1043-extension-members-for-operators-and-srtp-constraints.md
